src/loan/ExitQueuePSM.sol

```diff
@@ -0,0 +1,303 @@
+                amountRemaining: amount,
+                owner: msg.sender,
+                timestamp: uint64(timestamp),
+                feePercent: feePct
+            })
+        );
+
+        ExitQueueTicket storage storedTicket = _tickets[ticketId];
+
+        require(
+            storedTicket.timestamp != 0,
+            "ExitQueuePSM: cannot find ticket"
+        );
+
+        require(
+            storedTicket.amountRemaining != 0,
+            "ExitQueuePSM: no amount to withdraw"
+        );
+
+        // here we don't delete the ticket in the tickets mapping so that
+        // the delete can be done during the mint (to get gas refund to the minter)
+        uint256 amountToSend = storedTicket.amountRemaining;
+        storedTicket.amountRemaining = 0;
+        ERC20(credit).safeTransfer(msg.sender, amountToSend);
+    }
+
+    /// @notice Enters the exit queue with a ticket
+    /// @dev must have approved the PSM for the 'creditAmount'
+    /// @dev if feePct > 0, the feePct must be higher than the fee of the current front ticket
+    /// @param creditAmount The amount of CREDIT to enter into the queue
+    /// @param feePct The fee percentage for entering the queue. 0.01e18 = 1%
+    function enterExitQueue(uint256 creditAmount, uint64 feePct) public {
+        require(creditAmount >= MIN_AMOUNT, "ExitQueuePSM: amount too low");
+        require(feePct < 1e18, "ExitQueuePSM: fee should be < 100%");
+
+        ExitQueueTicket memory ticket = ExitQueueTicket({
+            amountRemaining: creditAmount,
+            owner: msg.sender,
+            timestamp: uint64(block.timestamp),
+            feePercent: feePct
+        });
+
+        bytes32 ticketId = _getTicketId(ticket);
+        require(
+            _tickets[ticketId].timestamp == 0,
+            "ExitQueuePSM: exit queue ticket already saved"
+        );
+
+        // transfer tokens to the PSM
+        ERC20(credit).safeTransferFrom(msg.sender, address(this), creditAmount);
+
+        // save the exit queue ticket to the tickets mapping
+        _tickets[ticketId] = ticket;
+
+        if (feePct == 0 || queue.empty()) {
+            // no fees or queue empty, push to the back of the queue
+            queue.pushBack(ticketId);
+        } else {
+            bytes32 frontTicketId = queue.front();
+            ExitQueueTicket memory frontTicket = _tickets[frontTicketId];
+            require(
+                frontTicket.feePercent < feePct,
+                "ExitQueuePSM: Can only outdiscount current high discounter"
+            );
+
+            queue.pushFront(ticketId);
+        }
+
+        emit EnterExitQueue(
+            msg.sender,
+            creditAmount,
+            feePct,
+            block.timestamp,
+            ticketId
+        );
+    }
+
+    /// @notice mint `amountOut` CREDIT to address `to` for a maximum of`amountIn` underlying tokens
+    /// It can cost less than amountIn if any tickets in the exit queue have discount fees > 0%
```

Comment:
I think that is weird, because `amountIn` is the input data of the function. With a proper `getMintAmountOut` implementation, the user knows it will need to input less peg tokens if they want, but the function should "consume" the full amountIn